### PR TITLE
Cr 1326 transient uac journey

### PR DESCRIPTION
**Pull request by Paul-Joel**
# Motivation and Context
New Transient UAC journey

# What has changed
Introduces new 'transient' UAC journey.
If estabType of entered UAC is set as 'Transient' the user is asked to enter their nearest Town, and then to select their accommodation type. These are then combined and fed to EQ in the 'display_address' value.
Code has normal error trapping and required unit tests

No Cucumber updates are required to merge, although specific tests may need building

# How to test?
'Create' a transient UAC, and enter it's UAC into /en/start/
Confirm get new journey

# Links
https://collaborate2.ons.gov.uk/jira/browse/CR-1303

# Screenshots (if appropriate):

### CODE REVIEW

File: app/start_handlers.py [hunk 1]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+            case = session['case']
+        except KeyError:
+            raise SessionTimeout('start')
+
+        try:
+            accommodation_type = data['accommodation-type']
+            session['attributes']['transientAccommodationType'] = accommodation_type
+            session.changed()
+
+            if case['region'][0] == 'N':
```
Comment (cookel2) [inline app/start_handlers.py:811]: I'm not certain but I think you can remove the [0] here.

**Review by cookel2 — CHANGES_REQUESTED** [review]
It all worked nicely when I tested it. There was one small thing (see comment). Also, this might just be something to do with my local env, but running "make test" gave me this warning:
flake8 app
app/utils.py:85:5: E303 too many blank lines (3)
make: *** [flake8] Error 1

**Review by cookel2 — APPROVED** [review]
looks good

File: tests/test_data/rhsvc/uac_unlinked_n.json [hunk 2]
```diff
@@ -9,5 +9,6 @@
 	"collectionExerciseId": "4a6c6e0a-6384-4da8-8c3c-7c56a801f792",
 	"address": null,
 	"formType": "H",
+	"estabType": "HOUSEHOLD",
 	"handDelivery": false
 }
```
Comment (MartinHumphrey) [inline tests/test_data/rhsvc/uac_unlinked_n.json:14]: Add CR/LF.

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -309,3 +312,20 @@ def test_get_call_centre_number(self):
...
+            mocked_get_now_utc.return_value = mocked_now_utc
+            self.assertTrue(View.check_if_after_census_day())
+
+    def before_census_day(self, year, month=None, day=None, hour=0, minute=0, second=0):
+        mocked_now_utc = datetime.datetime(year, month, day, hour, minute, second, 0)
+        with mock.patch('app.utils.View.get_now_utc') as mocked_get_now_utc:
+            mocked_get_now_utc.return_value = mocked_now_utc
+            self.assertFalse(View.check_if_after_census_day())
+
+    def test_check_before_after_census_day(self):
```
Comment (MartinHumphrey) [inline tests/unit/test_utils.py:328]: Are any of these attributes that are set-up used anywhere? All changes to this file could be removed?

**Review by MartinHumphrey — CHANGES_REQUESTED** [review]
Minor changes.

File: tests/unit/test_utils.py (hunk 3 above)
Comment (Paul-Joel) [reply tests/unit/test_utils.py:328]: Don't understand the comment

File: tests/test_data/rhsvc/uac_unlinked_n.json (hunk 2 above)
Comment (Paul-Joel) [reply tests/test_data/rhsvc/uac_unlinked_n.json]: Done

File: tests/test_data/rhsvc/uac_unlinked_w.json [hunk 4]
```diff
@@ -9,5 +9,6 @@
 	"collectionExerciseId": "4a6c6e0a-6384-4da8-8c3c-7c56a801f792",
 	"address": null,
 	"formType": "H",
+	"estabType": "HOUSEHOLD",
 	"handDelivery": false
 }
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/rhsvc/uac_unlinked_w.json]: Done

File: tests/test_data/rhsvc/uac_transient_w.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+	"caseType": "SPG",
+	"estabType": "Transient",
+	"formType": "H",
+	"region": "W",
+	"caseId": "e37b0d05-3643-445e-8e71-73f7df3ff95e",
+	"collectionExerciseId": "22684ede-7d5f-4f53-9069-2398055c61b2",
+	"address": {
+		"uprn": "9999999999999"
+	}
+}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/rhsvc/uac_transient_w.json]: Done

File: tests/test_data/rhsvc/uac_transient_n.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+	"caseType": "SPG",
+	"estabType": "Transient",
+	"formType": "H",
+	"region": "N",
+	"caseId": "e37b0d05-3643-445e-8e71-73f7df3ff95e",
+	"collectionExerciseId": "22684ede-7d5f-4f53-9069-2398055c61b2",
+	"address": {
+		"uprn": "9999999999999"
+	}
+}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/rhsvc/uac_transient_n.json]: Done

File: tests/test_data/rhsvc/uac_transient_e.json [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+	"caseType": "SPG",
+	"estabType": "Transient",
+	"formType": "H",
+	"region": "E",
+	"caseId": "e37b0d05-3643-445e-8e71-73f7df3ff95e",
+	"collectionExerciseId": "22684ede-7d5f-4f53-9069-2398055c61b2",
+	"address": {
+		"uprn": "9999999999999"
+	}
+}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply tests/test_data/rhsvc/uac_transient_e.json]: Done

File: app/start_handlers.py [hunk 8]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+                attributes['language'] = locale
+                attributes['display_region'] = display_region
+                await self.call_questionnaire(request, case,
+                                              attributes, request.app,
+                                              session.get('adlocation'))
+
+        except KeyError:
+            logger.info('transient accommodation type error', client_ip=request['client_ip'])
+            if display_region == 'cy':
+                flash(request, FlashMessage.generate_flash_message('Select an answer', 'ERROR',
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:803]: Done

File: app/templates/start-transient-accommodation-type.html [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    {{
+        onsButton({
+            'text': 'Continue',
+            'classes': 'u-mt-l',
+            'name': 'action[save_continue]'
+        })
+    }}
+
+{%- endblock -%}
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/templates/start-transient-accommodation-type.html]: Done

File: app/start_handlers.py [hunk 10]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+
+        self.log_entry(request, display_region + '/start/transient/enter-town-name')
+
+        data = await request.post()
+
+        session = await get_session(request)
+
+        try:
+            town_name = data['form-enter-town-name']
+            if town_name == '':
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py]: Changed

File: app/start_handlers.py [hunk 11]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+            'page_show_signout': 'true'
+        }
+
+    @aiohttp_jinja2.template('start-transient-enter-town.html')
+    async def post(self, request):
+        self.setup_request(request)
+        await check_permission(request)
+        display_region = request.match_info['display_region']
+
+        if display_region == 'cy':
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:681]: Changed

File: app/start_handlers.py [hunk 12]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+
+            raise HTTPFound(
+                request.app.router['StartTransientAccommodationType:get'].url_for(display_region=display_region)
+            )
+
+        except KeyError:
+            logger.info('error town name empty',
+                        client_ip=request['client_ip'])
+            if display_region == 'cy':
+                flash(request, FlashMessage.generate_flash_message('Enter your nearest town or city', 'ERROR',
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py:718]: Done

File: app/start_handlers.py [hunk 13]
```diff
@@ -655,3 +659,179 @@ async def post(self, request):
...
+@start_routes.view(r'/' + View.valid_display_regions + '/start/transient/enter-town-name/')
+class StartTransientEnterTownName(StartCommon):
+    @aiohttp_jinja2.template('start-transient-enter-town.html')
+    async def get(self, request):
+        self.setup_request(request)
+        await check_permission(request)
+        display_region = request.match_info['display_region']
+
+        if display_region == 'cy':
+            if View.check_if_after_census_day():
```
[the remark this answers is not in the document]
Comment (Paul-Joel) [reply app/start_handlers.py]: Changed

File: tests/unit/test_utils.py (hunk 3 above)
Comment (MartinHumphrey) [reply tests/unit/test_utils.py:328]: All changes to test_utils.py appear to be redundant code. They are not called and could be removed with no effect on running the unit tests?

**Review by MartinHumphrey — APPROVED** [review]

**Pull request merged**